Cannot create shared libraries from D sources on OSX (and macOS)

Using the `library` function in a 'meson.build' to create a shared library on Debian Sid and Fedora Rawhide works fine. For example:
```
tinyendian_lib = library(
    meson.project_name(),
    [tinyendian_src],
    include_directories: [src_dir],
    version: meson.project_version(),
    pic: true,
    install: true,
)
```
correctly builds the '.so' file and it all installs fine. However on OSX (I am on El Capitan) and I assume macOS, the option `-L-soname` is passed to the 'ldc2' command line, and whilst this works for Linux, it does not work for OSX:
```
  FAILED: libtinyendian.0.dylib 
ldc2  -of libtinyendian.0.dylib 'tinyendian@sha/source_tinyendian.d.o' -shared -relocation-model=pic -L-soname -Llibtinyendian.dylib.0  
ld: unknown option: -soname
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: /usr/bin/gcc failed with status: 1
```
I believe this to be an error in the D tooling for Meson.

I fetched goalitium/d-soname and rebuilt the project on my El Capitan machine and:
```
|> ninja
[0/1] Regenerating build files.
The Meson build system
Version: 0.48.0.dev1
Source dir: /home/users/russel/Repositories/Git/Forks/TinyEndian
Build dir: /Users/russel/BuildArea/TinyEndian
Build type: native build
Project name: tinyendian
Project version: 0.2.0
Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0git-16f3746):")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Build targets in project: 1
Found ninja-1.8.2 at /usr/local/bin/ninja
[2/2] Linking target libtinyendian.0.dylib.
FAILED: libtinyendian.0.dylib 
ldc2  -of libtinyendian.0.dylib 'tinyendian@sha/source_tinyendian.d.o' -shared -relocation-model=pic -install_name @rpath/libtinyendian.0.dylib  
ldc2: Unknown command line argument '-install_name'.  Try: 'ldc2 -help'
ldc2: Did you mean '-soname'?
ninja: build stopped: subcommand failed.
```

--- a/mesonbuild/compilers/d.py
+++ b/mesonbuild/compilers/d.py
@@ -221,6 +221,9 @@ class DCompiler(Compiler):
                 for la in linkargs:
                     dcargs.append('-L' + la.strip())
                 continue
+            elif arg.startswith('-install-name'):
+                dcargs.append('-L' + arg)
+                continue
             elif arg.startswith('-link-defaultlib') or arg.startswith('-linker'):
                 # these are special arguments to the LDC linker call,
                 # arguments like "-link-defaultlib-shared" do *not*

--- a/mesonbuild/compilers/d.py
+++ b/mesonbuild/compilers/d.py
@@ -14,10 +14,12 @@
 
 import os.path, subprocess
 
-from ..mesonlib import EnvironmentException, version_compare
+from ..mesonlib import EnvironmentException, version_compare, is_windows, is_osx
 
 from .compilers import (
     GCC_STANDARD,
+    GCC_CYGWIN,
+    GCC_OSX,
     d_dmd_buildtype_args,
     d_gdc_buildtype_args,
     d_ldc_buildtype_args,
@@ -90,8 +92,15 @@ class DCompiler(Compiler):
         return ['-shared']
 
     def get_soname_args(self, prefix, shlib_name, suffix, soversion, is_shared_module):
-        # FIXME: Make this work for Windows, MacOS and cross-compiling
-        return get_gcc_soname_args(GCC_STANDARD, prefix, shlib_name, suffix, soversion, is_shared_module)
+        # FIXME: Make this work for cross-compiling
+        gcc_type = GCC_STANDARD
+        if is_windows():
+            gcc_type = GCC_CYGWIN
+        if is_osx():
+            gcc_type = GCC_OSX
+
+        return get_gcc_soname_args(gcc_type, prefix, shlib_name, suffix, soversion, is_shared_module)
+
 
     def get_feature_args(self, kwargs, build_to_src):
         res = []

